pkg/portal/service/app.go

```diff
@@ -113,10 +113,10 @@ func (s *AppService) Get(id string) (*model.App, error) {
 	}, nil
 }
 
-func (s *AppService) GetAppProvider(appID string) (*deps.AppProvider, error) {
+func (s *AppService) WithAppProvider(appID string, executor func(*deps.AppProvider) error) error {
```

Comment:
I prefer we introduce a new event service to log events, instead of reusing the original one in pkg/lib/event. The new event service installs the hook in the database handle being used in the portal graphql handler. So the bug will not occur in the first place. The new event service only needs the audit sink, skip other irrelevant sinks.

Reply:
@louischan-oursky Actually we still need another transaction for writing to app audit db because it might be another db?

Reply:
The new event service manages the transaction within itself. Note that in pkg/portal/transport/graphql_handler.go, a read only transaction is opened there. May need to adjust transaction isolation level to avoid deadlock.

Reply:
@louischan-oursky In the original event service, we have a user resolver so we can resolve user refs in the service. To implement that resolver we need a lot of things inside the AppProvider.
If we make a new event service, I think we don't implement a new resolver for only this service. So user ref resolving will not work in this service.
Currently, all portal mutations doesn't contains user ref in the payload so this still works so I can do in this way first. However we need to implement a new resolver if we want to resolve user refs in portal events in the future.